spring-cloud-gateway-core/src/main/java/org/springframework/cloud/gateway/filter/headers/XForwardedHeadersFilter.java

```diff
@@ -187,6 +219,37 @@ public HttpHeaders filter(HttpHeaders input, ServerWebExchange exchange) {
 			write(updated, X_FORWARDED_PROTO_HEADER, proto, isProtoAppend());
 		}
 
+		if(isPrefixEnabled()) {
+			//if the path of the url that the gw is routing to is a subset (and ending part) of the url that it is routing from then the difference is the prefix
+			//e.g. if request original.com/prefix/get/ is routed to routedservice:8090/get then /prefix is the prefix - see XForwardedHeadersFilterTests
+			//so first get uris, then extract paths and remove one from another if it's the ending part
+
+			LinkedHashSet<URI> originalUris = exchange.getAttribute(GATEWAY_ORIGINAL_REQUEST_URL_ATTR);
+			URI requestUri = exchange.getAttribute(GATEWAY_REQUEST_URL_ATTR);
+
+			if(originalUris != null && requestUri != null) {
+
+				originalUris.stream().forEach(originalUri -> {
+
+					if(originalUri!=null && originalUri.getPath()!=null) {
+						String prefix = originalUri.getPath();
+
+						//strip trailing slashes before checking if request path is end of original path
+						String originalUriPath = originalUri.getPath().substring(0, originalUri.getPath().length() - (originalUri.getPath().endsWith("/") ? 1 : 0));
```

Comment:
No need to do a substring if path doesn't end with `/`. Refactor to something like
```
	private String stripTrailingSlash(URI uri) {
		if (uri.getPath().endsWith("/")) {
			return uri.getPath().substring(0, uri.getPath().length() - 1);
		} else {
			return uri.getPath();
		}
	}
```